plugins/out_stackdriver/stackdriver.c

```diff
@@ -361,152 +361,172 @@ static struct mk_list *parse_local_resource_id_to_list(char *str, char *type)
-            msgpack_unpacked_destroy(&result);
-            return -1;
+        if (!ctx->local_resource_id) {
+            /* if local_resource_id is not found, use the tag of the log */
+            flb_plg_debug(ctx->ins, "local_resource_id not found, "
+                                    "tag [%s] is assigned for local_resource_id", tag);
+            ctx->local_resource_id = flb_sds_create(tag);
         }
+    }
+    else {
+        msgpack_unpacked_destroy(&result);
+        flb_plg_error(ctx->ins, "failed to unpack data");
+        return -1;
+    }
 
-        if (strncmp(type, K8S_CONTAINER, len_k8s_container) == 0) {
-            list = parse_local_resource_id_to_list(local_resource_id, K8S_CONTAINER);
-            if (!list) {
-                goto error;
-            }
+    msgpack_unpacked_destroy(&result);
+    return 0;
+}
 
-            /* iterate through the list */
-            mk_list_foreach_safe(head, tmp, list) {
-                ptr = mk_list_entry(head, struct local_resource_id_list, _head);
-                if (first) {
-                    /* check the prefix */
-                    if (flb_sds_len(ptr->val) != len_k8s_container ||
-                        strncmp(ptr->val, K8S_CONTAINER, len_k8s_container) != 0) {
-                        goto error;
-                    }
-                    first = 0;
-                    continue;
-                }
+/* 
+ *    process_local_resource_id():
+ *  - use the extracted local_resource_id to assign the label keys for different
+ *    resource types that are specified in the configuration of stackdriver_out plugin
+ */
+static int process_local_resource_id(struct flb_stackdriver *ctx, char *type)
+{
+    int ret = -1;
+    int first = 1;
+    int len_k8s_container;
+    int len_k8s_node;
+    int len_k8s_pod;
+    struct local_resource_id_list *ptr;
+    struct mk_list *list = NULL;
+    struct mk_list *tmp;
+    struct mk_list *head;
 
-                /* Follow the order of fields in local_resource_id */
-                if (!ctx->namespace_name) {
-                    ctx->namespace_name = flb_sds_create(ptr->val);
-                }
-                else if (!ctx->pod_name) {
-                    ctx->pod_name = flb_sds_create(ptr->val);
-                }
-                else if (!ctx->container_name) {
-                    ctx->container_name = flb_sds_create(ptr->val);
+    if (!ctx->local_resource_id) {
+        flb_plg_error(ctx->ins, "local_resource_is is not assigned");
+        return -1;
+    }
+
+    len_k8s_container = sizeof(K8S_CONTAINER) - 1;
+    len_k8s_node = sizeof(K8S_NODE) - 1;
+    len_k8s_pod = sizeof(K8S_POD) - 1;
+
+    if (strncmp(type, K8S_CONTAINER, len_k8s_container) == 0) {
+        list = parse_local_resource_id_to_list(ctx->local_resource_id, K8S_CONTAINER);
+        if (!list) {
+            goto error;
+        }
+
+        /* iterate through the list */
+        mk_list_foreach_safe(head, tmp, list) {
```

Comment:
```mk_list_foreach_safe()``` is primary used when one of the nodes of the list while iterating might be deleted. If is just a read-only operation you have to use ```mk_list_foreach()```

Reply:
updated